library/src/contraction/contraction_solution_impl.hpp

```diff
@@ -232,10 +389,723 @@ namespace hiptensor
+            B_d_real = std::move(allocDevice<elementType>(elementsB));
+            B_d_imag.reset(nullptr);
+            B_d_imag = std::move(allocDevice<elementType>(elementsB));
+            D_d_real.reset(nullptr);
+            D_d_real = std::move(allocDevice<elementType>(elementsD));
+            D_d_imag.reset(nullptr);
+            D_d_imag = std::move(allocDevice<elementType>(elementsD));
+            E_d_real.reset(nullptr);
+            E_d_real = std::move(allocDevice<elementType>(elementsE));
+            E_d_imag.reset(nullptr);
+            E_d_imag = std::move(allocDevice<elementType>(elementsE));
+
+            auto blockDim = dim3(1024, 1, 1);
+            auto gridDim  = dim3(ceilDiv(elementsA, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)A,
+                            (elementType*)A_d_real.get(), (elementType*)A_d_imag.get(), elementsA);
+
+            gridDim  = dim3(ceilDiv(elementsB, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)B,
+                            (elementType*)B_d_real.get(), (elementType*)B_d_imag.get(), elementsB);
+
+            gridDim  = dim3(ceilDiv(elementsD, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)D,
+                            (elementType*)D_d_real.get(), (elementType*)D_d_imag.get(), elementsD);
+
+            gridDim  = dim3(ceilDiv(elementsE, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)E,
+                            (elementType*)E_d_real.get(), (elementType*)E_d_imag.get(), elementsE);
+
+            // Initialize the argument pointer
+            mArgPtr[0] = std::move(deviceOp->MakeArgumentPointer(
+                A_d_real.get(),
+                B_d_real.get(),
+                std::array<const void*, 1>{D_d_real.get()},
+                E_d_real.get(),
+                toCKVec(a_ms_ks_lengths),
+                toCKVec(a_ms_ks_strides),
+                toCKVec(b_ns_ks_lengths),
+                toCKVec(b_ns_ks_strides),
+                std::array<std::vector<ck::index_t>, 1>{toCKVec(ds_ms_ns_lengths)},
+                std::array<std::vector<ck::index_t>, 1>{toCKVec(ds_ms_ns_strides)},
+                toCKVec(e_ms_ns_lengths),
+                toCKVec(e_ms_ns_strides),
+                typename Traits::AOp{},
+                typename Traits::BOp{},
+                typename Traits::CDEOp{alphaF, betaF}));
+
+            // Attach the workspace pointer
+            deviceOp->SetWorkSpacePointer(mArgPtr[0].get(), workspacePtr);
+
+            // Initialize the invoker
+            mInvokerPtr[0] = std::move(deviceOp->MakeInvokerPointer());
```

Comment:
Not sure that there is a need for 4 separate invokers because they are all the same kernel. We can try to feed each of the separate args kernels to just a single invoker ptr.